### ISSUE TITLE
[TODO #4]: Implement backpropagation and training loop

### ISSUE THREAD

**Issue by SukritTM** [The5thAxiom/parallel-neural-networks#7]
- [ ] Implement backpropagation function for each layer
    - function inputs: gradient of cost w.r.t layer output
    - function effects: update weights by calculating gradient of cost w.r.t weights
    - function returns: gradient of cost w.r.t layer input
- [ ] Implement training loop
    - Each train step, the gradient of the output layer is modified by calling 'backprop' for that layer
    - the output from that is passed as input to the previous layer's 'backprop' method
    - This is done until the end of the layers
